feat: add environment variable support for client initialization params

LilDojd opened the pull request:
The commit adds support for initializing AlchemiscaleClient with environment variables, allowing configuration via ALCHEMISCALE_URL, ALCHEMISCALE_ID, and ALCHEMISCALE_KEY

Adds test for this behavior

Closes #355

LilDojd commented:
Not sure where to update docs for this

dotsdl commented:
Awesome, thank you @LilDojd! Will review this as part of our [0.6.1](https://github.com/OpenFreeEnergy/alchemiscale/milestone/18) release cycle coming up next!

ianmkenney reviewed (changes requested):
Thanks for the PR @LilDojd! I like the overall design you've implemented and have left some comments/suggestions.

ianmkenney commented:
If you could also update the docs, that would be appreciated!

I think the best place for that would be around here:

https://github.com/OpenFreeEnergy/alchemiscale/blob/d2ab829f6f462b17f4ef20108b4e8478f6716f8d/docs/user_guide.rst?plain=1#L88

LilDojd commented on alchemiscale/base/client.py:
```diff
@@ -52,17 +54,97 @@ async def _wrapper(self, *args, **kwargs):
+        ValueError
+            If neither param_value nor environment variable is set
+        """
+        env_value = os.getenv(self.env_var_name)
+
+        if param_value is None:
+            param_value = env_value
+        elif env_value is not None and param_value != env_value:
+            self._warn_override(param_value, env_value)
+
+        if param_value is None:
+            raise ValueError(
+                f"No {self.human_name} provided and {self.env_var_name} environment variable not set"
+            )
+
+        return param_value
```
Too bad the pyright does not do type narrowing for tuple expressions, so it is a bit awkward with type checkers in `(param_value, env_value)` case

LilDojd commented on alchemiscale/base/client.py:
```diff
@@ -52,17 +54,97 @@ async def _wrapper(self, *args, **kwargs):
+    param_name: str
+    env_var_name: str
+    human_name: str
+    render_value: bool = True
```
No, not really. It was `censor_value` when I wrote it first, so I probably just forgot to change the default to False

LilDojd commented:
@ianmkenney Thank you for your review!
I believe I have addressed all your comments and suggestions, and updated the docs

@dotsdl Let me know if you would prefer to use pydantic for this one.  I could draft another PR candidate with `pydantic.BaseSettings` like they have in [asap-discovery](https://github.com/asapdiscovery/asapdiscovery/blob/21a538b51f179e642df27bb9f7af16fe13dc7d8b/asapdiscovery-alchemy/asapdiscovery/alchemy/interfaces.py#L4)

ianmkenney reviewed (changes requested):
This looks really good. Just some notes on developer-specific tooling comments.

dotsdl reviewed (approved):
Looks great @LilDojd! Thanks @ianmkenney for helping to improve this PR! Feel free to merge when satisfied!
